### PR TITLE
accounts/scwallet: add support for smartcard wallets

### PR DESCRIPTION
**Pull request by Arachnid**
Details on interaction with the smartcard can be found in wallets/scwallet/hub.go.

Compatible smart cards can be purchased here: https://www.smartcardfocus.com/shop/ilp/id~790/acosj-dual-interface-java-card/p/index.shtml. Any USB reader will suffice; they all use the same standard.

In place of a real card, a virtual card can be used by running [jcardsim](https://github.com/status-im/jcardsim).

If you buy a blank card or use a simulator, you will need to build and deploy the applet, as documented [here](https://github.com/status-im/hardware-wallet/blob/master/README.md).

### CODE REVIEW

File: accounts/scwallet/hub.go [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+// refreshCycle is the maximum time between wallet refreshes (if USB hotplug
+// notifications don't work).
+const refreshCycle = 5 * time.Second
+
+// refreshThrottling is the minimum time between wallet refreshes to avoid thrashing.
+const refreshThrottling = 500 * time.Millisecond
+
+// SmartcardPairing contains information about a smart card we have paired with
+// or might pair withub.
```
Comment (karalabe) [inline accounts/scwallet/hub.go]: `withub`?

File: accounts/scwallet/hub.go [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+type SmartcardPairing struct {
```
Comment (karalabe) [inline accounts/scwallet/hub.go]: Any reason for having this a public type?

File: accounts/scwallet/hub.go [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+	for _, wallet := range hub.wallets {
+		if wallet != nil {
+			cpy = append(cpy, wallet)
+		}
+	}
+	return cpy
+}
+
+// refreshWallets scans the USB devices attached to the machine and updates the
+// list of wallets based on the found devices.
```
Comment (karalabe) [inline accounts/scwallet/hub.go]: USB?

File: accounts/scwallet/hub.go [hunk 4]
```diff
@@ -0,0 +1,277 @@
...
+
+	if err := hub.readPairings(); err != nil {
+		return nil, err
+	}
+
+	hub.refreshWallets()
+	return hub, nil
+}
+
+// Wallets implements accounts.Backend, returning all the currently tracked USB
```
Comment (karalabe) [inline accounts/scwallet/hub.go]: USB?

File: node/config.go [hunk 5]
```diff
@@ -423,6 +424,12 @@ func makeAccountManager(conf *Config) (*accounts.Manager, string, error) {
 		} else {
 			backends = append(backends, trezorhub)
 		}
+		// Start a smart card hub
+		if schub, err := scwallet.NewHub(scwallet.Scheme, keydir); err != nil {
```
[the remark this answers is not in the document]
Comment (Arachnid) [reply node/config.go]: For now, yes. Hardcoding it seemed like a bad idea.

**Comment by Arachnid** [comment]
> There's also a new dependency on libpcsclite.pc. This might be a biggie to solve, for USB support I explicitly made a library that contains all the C code we need too so we don't need the users installing any extra dev libs. Not sure how best to solve this.

Could we enable smartcard wallets via flag only, and document that they will need pcsclite installed if they want smartcard support?

**Comment by karalabe** [comment]
That would be okish, but currently geth won't build without it.

**Comment by hrishikeshio** [comment]
I am getting following error:
```
> personal.openWallet('pcsc://496d','123456789012')
Error: Invalid card cryptogram
```
@Arachnid

**Comment by gballet** [comment]
Superseded by #19273

**Pull request closed without merging**